### DISCUSSION TITLE
Noob Issues

### DISCUSSION THREAD

**Discussion by amurbain** [seanap/auto-m4b#29] [General]
I'm having difficulties getting this to work. To lay the groundwork:

Plex is hosted on Windows. A "working directory" has been created that houses the working directories, this is a network share. 

Running the `auto-m4b` docker image in a Ubuntu Server 22.0.1 LTS VM in VMware Workstation. The working directory is accessible to to the Linux VM. 

I have placed an audiobook's worth of MP3s in the `recentlyadded` folder and have waited much longer than 5 minutes and have checked the `backup`, `merge`, and `untagged` folders for any output and have found nothing.

During the configuration process, I noticed in the `docker-compose.yml` it refers to `/path/to/config:/config`, I'm not sure what config is supposed to be referenced here, so I omitted the line and only input my `/path/to/temp:/temp` path in as `/home/autom4b/temp:/temp` where `/temp` is mounted to the Windows share.

I was able to pull the log, see below:

![image](https://user-images.githubusercontent.com/1357207/235316592-ce24a44a-d3ac-46e9-bcc6-1c6ef672d659.png)

What am I doing wrong?

Thanks!

**Comment by seanap** [comment]
The config folder is where all the app data is stored and is very important.  It doesn't matter where you store your config folder but you need to have one.  

The way docker works is from the perspective of the docker container you map an external folder to an internal folder.  Kinda like an alias.  so when you see `/path/to/config:/config` Everything before the `:` is your folder on the Ubuntu OS (or NAS) and everything after the `:` is how the docker container references that folder inside the container and you shouldn't ever change.

Example:
`/usr/docker/autom4b/config:/config`

Many people mount their NAS to something like `/srv/storage`  and in this example to mount a temp folder that is inside your NAS would look something like this `/srv/storage/temp:/temp` .  I've never seen anyone mount their NAS to /home/autom4b/temp so make sure that is correct.

**Reply by DoloresHA** [reply]
Thank you- I didn't know what the config directory was "supposed" to be.

**Comment by amurbain** [comment]
I was just mirroring the folder structure from the readme (this folder structure is on the Windows share):

```
temp
│
└───recentlyadded # Input folder Add new books here
│     │     book1.m4b
│     |     book2.mp3
|     └─────book3
│           │   01-book3.mp3
│           │   ... 
└───merge # folder the script uses to combine mp3's
│     └─────book2
│           │   01-book2.mp3
│           │   ...
└───untagged # Output folder where all m4b's wait to be tagged
│     └─────book4
│           │   book4.m4b
└───delete # needed by the script
|
└───fix # Manually fix books with nested folders
|
└───backup # Backups incase anything goes wrong
      └─────book2
            │   01-book2.mp3
            │   ... 
```

Then I created the `temp` folder as instructed, which I made `/home/autom4b/temp`
I mounted the Windows share using 
`sudo mount -t cifs -o credentials=/etc/win-credentials //WIN_SHARE_IP/<share_name> /home/autom4b/temp`

I even went a step further and created a credentials file and added it to `fstab` so it'd auto-mount.
`//WIN_SHARE_IP/share_name  /home/autom4b/temp  cifs  credentials=/etc/win-credentials,file_mode=0755,dir_mode=0755 0 0`

When I browse to `/home/autom4b/temp` I can see the contents of the Windows share.

As for the config, I'm still not grasping exactly what is supposed to be in there. Just add the line and leave it be? I'm not storing anything on the container, it's just touching files on a Windows share (I obviously know nothing about Docker, sorry).

**Reply by seanap** [reply]
Ok that's good, your windows share looks just fine. 

The config folder is where all persistent data is stored.  You don't need to put anything in that folder but you have to put it in the docker compose file so the container can install necessary components to that folder location. You can put this in that line `/home/user/appdata/autom4b/config:/config` just replace user with your linux username

https://docs.linuxserver.io/general/running-our-containers#the-config-volume

Are the mp3's in a folder?

**Reply by amurbain** [reply]
Yes, it's currently
```
temp
│
└───recentlyadded - contains one folder with mp3s for testing
|     └─────book1
│           │   01-book1.mp3
│           │   ... 
└───merge - empty
|
└───untagged - empty
|
└───delete - empty
|
└───fix - empty
|
└───backup - empty
```

Some new info now showing when I `docker logs auto-m4b`:
![image](https://user-images.githubusercontent.com/1357207/235567290-2cecc049-d685-4068-bd0f-c2fcc1422e81.png)

**Comment by slinkoff** [comment]
did you ever get this to work? I have a similar issue in that auto-m4b doesn't detect anything in /temp/recentlyadded using Docker Desktop on Windows. 

Folder structure is definitely correct. If i manually run the line from the script:

`find /temp/recentlyadded/ -maxdepth 2 -mindepth 2 -type f \( -name '*.mp3' -o -name '*.m4b' -o -name '*.m4a' \) -print0 | xargs -0 -L 1 dirname | sort | uniq -c | grep -E -v '^ *1 ' | sed 's/^ *[0-9]* //' | while read i; do mv -v "$i" /temp/merge/; done`

It finds and moves the files and then I get a different error:

```
2023-10-30 12:41:12 Starting Conversion
2023-10-30 12:41:12 an error occured, that has not been caught:
2023-10-30 12:41:12 Array
2023-10-30 12:41:12 (
2023-10-30 12:41:12     [type] => 1
2023-10-30 12:41:12     [message] => Uncaught ValueError: Path cannot be empty in phar:///usr/local/bin/m4b-tool/src/library/M4bTool/Command/AbstractCommand.php:228
2023-10-30 12:41:12 Stack trace:
2023-10-30 12:41:12 #0 phar:///usr/local/bin/m4b-tool/src/library/M4bTool/Command/AbstractCommand.php(228): file_put_contents()
2023-10-30 12:41:12 #1 phar:///usr/local/bin/m4b-tool/src/library/M4bTool/Command/AbstractCommand.php(462): M4bTool\Command\AbstractCommand->log()
2023-10-30 12:41:12 #2 phar:///usr/local/bin/m4b-tool/src/library/M4bTool/Command/MergeCommand.php(197): M4bTool\Command\AbstractCommand->error()
2023-10-30 12:41:12 #3 phar:///usr/local/bin/m4b-tool/vendor/symfony/console/Command/Command.php(255): M4bTool\Command\MergeCommand->execute()
2023-10-30 12:41:12 #4 phar:///usr/local/bin/m4b-tool/vendor/symfony/console/Application.php(908): Symfony\Component\Console\Command\Command->run()
2023-10-30 12:41:12 #5 phar:///usr/local/bin/m4b-tool/vendor/symfony/console/Application.php(269): Symfony\Component\Console\Application->doRunCommand()
2023-10-30 12:41:12 #6 phar:///usr/local/bin/m4b-tool/vendor/symfony/console/Application.php(145): Symfony\Component\Console\Application->doRun()
2023-10-30 12:41:12 #7 phar:///usr/local/bin/m4b-tool/bin/m4b-tool.php(32): Symfony\Component\Console\Application->run()
2023-10-30 12:41:12 #8 /usr/local/bin/m4b-tool(10): require('...')
2023-10-30 12:41:12 #9 {main}
2023-10-30 12:41:12   thrown
2023-10-30 12:41:12     [file] => phar:///usr/local/bin/m4b-tool/src/library/M4bTool/Command/AbstractCommand.php
2023-10-30 12:41:12     [line] => 228
2023-10-30 12:41:12 )
2023-10-30 12:41:12 Finished Converting
2023-10-30 12:41:12 Putting the m4b into a folder
```

This just loops round continuously, it hasn't actually finished converting anything. I have to move the files back to /temp/recentlyadded to get it to stop

docker compose YAML:

```
version: '3.7'
services:
  auto-m4b:
    image: seanap/auto-m4b:latest
    container_name: auto-m4b
    volumes:
      - "M:/Downloads/M4B:/temp"
      - "m4b-config:/config"
    environment:
      - PUID=1000
      - PGID=1000
      - CPU_CORES=2
      - SLEEPTIME=1m
      - MAKE_BACKUP=N
     
volumes:
  m4b-config:
```

**Comment by slinkoff** [comment]
ok, this is interesting. I tried running auto-m4b-tool.sh from inside the container and it worked fine. I noticed edits I made to the file didn't change what was appearing in the log, so what is the cron job running? How is the script being executed? Not very familiar with the workings of Linux to figure it out but if it was calling the auto-m4b-tool.sh in the root folder then it should work, but it isn't
